File: javascript/two_sum.js
var twoSum = function(nums, target) {
   let memory = new Map();      // memory' will store met numbers along with indexes to prevent iterating multiple times
   for (let i = 0; i < nums.length; i++) {      // for every possible nums index i
	   if (memory.has(target - nums[i])) {      // if seeking number has already been visited and is stored in memory
		   return [memory.get(target-nums[i]), i];      // just return an index of this number along with current index
	   }
	   else {
		   memory.set(nums[i], i);      // otherwise add the number to memory along with index
	   }
   }
	return [];      // in case the loop is exited, just return an empty array
};

File: python/two_sum.py
class Solution:
    def twoSum(self, nums, target):

        # Numbers that are needed to meet the target will be stored here along with an index of a complementary number.
        wanted_nums = {}

        # Interating through numbers list
        for i in range(len(nums)):

            # If number in wanted_nums it means we've got the sum!
            if nums[i] in wanted_nums:
                return [wanted_nums[nums[i]], i]

            # If not, we store the difference (so the number we seek) along with an index
            else:
                wanted_nums[target - nums[i]] = i

"""
This way the number of iterations is the smallest because all wanted numbers are stored in a memory.

I hope I helped you in some way. If you like this solution, I'll be very happy if you click "Star" on my Github page:
https://github.com/JBielan/leetcode-python

Thank you!
"""